test/unit/math/mix/fun/cholesky_decompose_test.cpp

```diff
@@ -8,59 +8,156 @@ namespace cholesky_decompose_test {
 // dof must be (n choose 2) + n
 auto f(int dof) {
   return [=](const auto& x) {
-    auto y = stan::math::cov_matrix_constrain(x, dof);
+    stan::math::promote_scalar_t<stan::value_type_t<decltype(x)>,
+                                 Eigen::Matrix<double, -1, -1>>
+        y;
+    try {
+      y = stan::math::cov_matrix_constrain(x, dof);
+    } catch (...) {
+      ADD_FAILURE() << "FAILED AT COV_MATRIX_CONSTRAIN";
+      throw;
+    }
     return stan::math::cholesky_decompose(y);
   };
 }
 
+auto f_matvar = [](const auto& x) { return stan::math::cholesky_decompose(x); };
+
 void expect_cholesky(const Eigen::MatrixXd& Sigma) {
   Eigen::VectorXd yy = stan::math::cov_matrix_free(Sigma);
-  stan::test::expect_ad(f(10), yy);
+  // lazy, solving for x in x = (N * (N + 1)) / 2
+  int dof = .5 * (std::sqrt(8 * yy.size() + 1) - 1);
+  stan::test::expect_ad(f(dof), yy);
+}
+template <typename F, typename T>
+void expect_cholesky_var(stan::test::ad_tolerances& tols, const F& f,
```

Comment:
@bbbales2 do you know anyway we can speed up this test? Right now I'm just checking the lower triangular of the output and it's still taking 5 ever

Reply:
How big are the matrices here?

Reply:
96 x 96 and 160 x 160 lol

Reply:
The issue is that it requires `25600` calls to grad and cholesky decompose for the full matrix. Or am I reading this wrong?

Reply:
Woof, yeah for reverse mode. Lol I didn't think about this but this is gonna be a bit of a trick.

Let me write up some code in R to do this real quick.

Reply:
Alright what I think we're going to have to do is instead of computing the Jacobian of this (which will be (NxN)x(NxN)), we gotta do the trick where we element-wise multiply the output of the function with some random thing and take the sum to reduce it to a scalar and look at the Jacobian there (which is just NxN).

It's not checking the full Jacobian but with the randomness we're still doing a good check (it'll be really hard to be accidentally right, at least):

So in pseudo-code the calculation is:
```
L = chol(A)
out = sum(L * matrix(runif(N * N), nrow = N))
out.grad()
// Check adjoints of A
```

There's a short analytic form for the adjoints we need in eq. 6 and 9 here: https://homepages.inf.ed.ac.uk/imurray2/pub/16choldiff/choldiff.pdf

Here is R code for that:
```
lchol = function(A) {
  t(chol(A))
}

Sigma = function(A) {
  out = matrix(0, nrow = nrow(A), ncol = ncol(A))
  for(i in 1:nrow(A)) {
    for(j in 1:i) {
      if(j == i)
        out[i, j] = A[i, j] / 2.0
      else
        out[i, j] = A[i, j]
    }
  }
  return(out)
}

N = 3
A = matrix(runif(N * N), nrow = N)
A = A + t(A)
A = A + 2.1 * N * diag(N) # 2 * N should be safe, 2.1 * N to be super safe

L = lchol(A)

# Use sigma here just to make lower triangular random matrix
adjL = Sigma(matrix(runif(N * N), nrow = N))

SigmaLinv = t(solve(t(L), t(Sigma(t(L) %*% adjL))))
S = solve(t(L), SigmaLinv)

# I did this a little differently so our adjoints would have an upper triangle of zero which
# I think matches current Stan behavior. Check this.
adjA = Sigma(S + t(S))
```

And here is finite difference code you can use to convince yourself that this R code is doing the right thing:

```
#### Compute a reference using finite difference (not needed for C++)
fd_adjA = matrix(0, nrow = N, ncol = N)
for(i in 1:N) {
  for(j in 1:i) {
    dx = 1e-3
    Ap = A
    Am = A
    Ap[i, j] = Ap[i, j] + dx
    Am[i, j] = Am[i, j] - dx
    Ap[j, i] = Ap[i, j]
    Am[j, i] = Am[i, j]
    
    adj = (sum(adjL * (lchol(Ap))) - sum(adjL * lchol(Am))) / (2 * dx)
    
    fd_adjA[i, j] = adj
  }
}

adjA
fd_adjA
```

Reply:
Oh. Whoops. That thing I'm calling a Sigma is a Phi in the paper lol. My bad lol.

Reply:
Actually Ben could we just bring back a bunch of the old rev tests? It looks like they are doing this?

https://github.com/stan-dev/math/blob/feature/flto/test/unit/math/rev/mat/fun/cholesky_decompose_test.cpp

Reply:
(That's just a random branch that has the old tests in it)

Reply:
Yeah I'm fine with whatever gets it tested. Just as long as we're vaguely testing gradients of cholesky for matrices larger than 35x35 or whatever the cutoff is.

Reply:
Woof @bbbales2 excellent catch on this, was very wrong! Now correct!